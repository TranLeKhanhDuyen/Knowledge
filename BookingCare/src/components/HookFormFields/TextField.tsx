import { ComponentPropsWithRef } from 'react'
import styles from './text-field.module.css'
import classNames from 'classnames/bind'

const className = classNames.bind(styles)

export interface TextFieldProps extends ComponentPropsWithRef<'input'> {
  label?: string
  helperText?: string
}

export default function TextField({
  name,
  label,
  helperText,
  ...other
}: TextFieldProps) {
  return (
    <div className={className('field-container')}>
      {label && (
        <label className={className('field-label')} htmlFor={name}>
          {label}
        </label>
      )}
      <div className={className('text-field')}>
        <input
          {...other}
          className={className('base-input', {
            'base-input-state-error': !!helperText,
            'base-input-state-default': !helperText
          })}
        />
      </div>
      {!!helperText && (
        <p
          className={className('helper-text', {
            'helper-text-state-error': !!helperText
          })}
        >
          {helperText}
        </p>
      )}
    </div>
  )
}
