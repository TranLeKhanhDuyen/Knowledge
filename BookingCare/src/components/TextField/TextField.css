.textfield-input {
  outline: none;
}

.textfield-input:focus {
  border: 1px solid var(--turquoise-color);
}

.select {
  outline: none;
}

.select:focus {
  border: 1px solid var(--turquoise-color);
}
