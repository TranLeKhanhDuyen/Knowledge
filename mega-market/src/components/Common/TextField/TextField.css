.text-field {
  width: 100%;
  outline: none;
  background-color: transparent;
  font-size: var(--font-sm);
  border: 1px solid transparent;
  padding: 2px;
}

.input-container {
  width: 100%;
  display: flex;
  align-items: center;
  gap: 10px;
  background-color: var(--tertiary-color);
  cursor: pointer;
  border-radius: 10px;
}
