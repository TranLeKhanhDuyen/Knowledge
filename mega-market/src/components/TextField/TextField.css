.text-field {
  width: 100%;
  outline: none;
  background-color: transparent;
  font-size: var(--font-md);
  border: 1px solid transparent;
  padding: 2px;
}

.input-container {
  display: flex;
  align-items: center;
  gap: 10px;
  background-color: var(--tertiary-color);
  padding: 15px;
  cursor: pointer;
  border-radius: 10px;
}
