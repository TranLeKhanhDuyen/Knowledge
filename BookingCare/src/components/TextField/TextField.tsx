import './TextField.css'
interface ITextField {
  label: string;
  type: string;
  name: string;
  value?: string;
  options?: string[];
}

export interface IListTextField {
  items: ITextField[];
}

const TextField = ({ items }: IListTextField) => {
  return (
    <>
      {items.map((item, index) => (
        <div key={index} className="text-field-item">
          <label>{item.label}</label>
          {item.type === 'select' ? (
            <select className='select' id={item.name} name={item.name} value={item.value}>
              {item.options?.map((option, optionIndex) => (
                <option className='option' key={optionIndex} value={option}>{option}</option>
              ))}
            </select>
          ) : (
            <input className='textfield-input' type={item.type} id={item.name} name={item.name} value={item.value} />
          )}
        </div>
      ))}
    </>
  );
};
export default TextField;
