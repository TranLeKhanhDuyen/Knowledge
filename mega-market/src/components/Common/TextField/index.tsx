import { ChangeEvent } from 'react'
import IconSvg, { IconSvgProps } from '@components/Common/IconSvg'
import './TextField.css'

export interface ITextFieldProps extends React.HTMLProps<HTMLInputElement> {
  isShowLabel?: boolean
  label?: string
  additionalClass?: string
  value?: string
  placeholder?: string
  iconLeft?: IconSvgProps['name']
  iconRight?: IconSvgProps['name']
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
  validate?: (value: string) => string | undefined
  error?: string
}

const TextField = ({
  isShowLabel = false,
  iconLeft,
  iconRight,
  label,
  additionalClass,
  value,
  placeholder,
  onChange,
  validate,
  error,
  ...props
}: ITextFieldProps) => {
  const handleChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
    onChange?.(event)
  }
  const validationError = validate ? validate(value || '') : undefined

  return (
    <>
      {isShowLabel && (
        <label className={`label-input ${additionalClass}`}>{label}</label>
      )}

      <div className={`input-container ${additionalClass}`}>
        {iconLeft && <IconSvg name={iconLeft} />}
        <input
          className={`text-field ${!isShowLabel ? 'hide' : ''}`}
          value={value}
          placeholder={placeholder}
          onChange={handleChangeInput}
          {...props}
        />
        {iconRight && <IconSvg name={iconRight} />}
      </div>
      <div className='error-placeholder'>
        {(error || validationError) && (
          <p className='error-message-filed'>{error || validationError}</p>
        )}
      </div>
    </>
  )
}

export default TextField
