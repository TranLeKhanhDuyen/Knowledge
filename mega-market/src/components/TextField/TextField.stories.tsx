import type { Meta, StoryObj } from '@storybook/react'
// Components
import TextField from '@components/TextField'

export default {
  title: 'Components/TextField',
  component: TextField,
  tags: ['autodocs'],
  argTypes: {
    isShowLabel: {
      description: 'One boolean to test for open or close label of input.'
    },
    label: { description: 'Content of label.' },
    validate: { description: 'Validation of input.' },
    additionalClass: {
      description: 'Adds the additional class to the text filed.'
    },
    placeholder: { description: 'Content of text field.' }
  }
} as Meta

type Story = StoryObj<typeof TextField>

export const InputText: Story = {
  args: {
    additionalClass: 'input-search',
    placeholder: 'Search ...'
  }
}

export const InputSearch: Story = {
  ...InputText,
  args: {
    ...InputText.args,
    isShowLabel: true,
    label: 'Email',
    placeholder: 'Enter email',
    additionalClass: 'input-text',
    iconLeft: 'search',
    iconRight: 'list'
  }
}
