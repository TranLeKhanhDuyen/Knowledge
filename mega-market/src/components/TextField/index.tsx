import { FormEvent } from 'react'

import '@components/TextField/TextField.css'

interface ITextFieldProps {
  isShowLabel?: boolean
  label?: string
  additionalClass?: string
  value?: string
  placeholder?: string
  iconLeft?: string
  iconRight?: string
  onChange?: (value: string) => void
  validate?: (value: string) => string | undefined
}

const TextField = ({
  isShowLabel = false,
  iconLeft,
  iconRight,
  label,
  additionalClass,
  value,
  placeholder,
  onChange,
  validate
}: ITextFieldProps) => {
  const handleChangeInput = (event: FormEvent<HTMLInputElement>) => {
    onChange?.(event.currentTarget.value)
  }

  const error = validate ? validate(value || '') : undefined

  return (
    <>
      <label className={`label-input ${!isShowLabel && 'hide'}`}>{label}</label>

      <div className='input-container'>
        {iconLeft && <img src={iconLeft} alt='icon' />}
        <input
          type='text'
          className={`text-field ${additionalClass}`}
          value={value}
          placeholder={placeholder}
          onChange={handleChangeInput}
        />
        {iconRight && <img src={iconRight} alt='icon' />}
        {error && <p className='error-message'>{error}</p>}
      </div>
    </>
  )
}

export default TextField
