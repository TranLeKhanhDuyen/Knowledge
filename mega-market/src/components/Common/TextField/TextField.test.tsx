import { render, fireEvent } from '@testing-library/react'
import TextField, { ITextFieldProps } from '.'

describe('TextField', () => {
  const renderTextField = (props: Partial<ITextFieldProps> = {}) => {
    const defaultProps: ITextFieldProps = {
      label: 'Username',
      placeholder: 'Enter username',
      onChange: jest.fn(),
      validate: (value: string) =>
        value.length < 6 ? 'Username too short' : undefined,
      ...props
    }
    return render(<TextField {...defaultProps} />)
  }

  test('renders label when isShowLabel is true', () => {
    const { getByText } = renderTextField({ isShowLabel: true })
    expect(getByText('Username')).toBeInTheDocument()
  })

  test('hides label when isShowLabel is false', () => {
    const { queryByText } = renderTextField({ isShowLabel: false })
    expect(queryByText('Username')).not.toBeInTheDocument()
  })

  test('renders input with placeholder', () => {
    const { getByPlaceholderText } = renderTextField()
    expect(getByPlaceholderText('Enter username')).toBeInTheDocument()
  })

  test('calls onChange callback when input value changes', () => {
    const onChangeMock = jest.fn()
    const { getByPlaceholderText } = renderTextField({ onChange: onChangeMock })
    const inputElement = getByPlaceholderText('Enter username')
    fireEvent.change(inputElement, { target: { value: 'user123' } })
    expect(onChangeMock).toHaveBeenCalledWith('user123')
  })

  test('displays error message if input is invalid', () => {
    const { getByText, getByPlaceholderText } = renderTextField()
    const inputElement = getByPlaceholderText('Enter username')
    fireEvent.change(inputElement, { target: { value: 'user' } })
    expect(getByText('Username too short')).toBeInTheDocument()
  })

  test('matches snapshot', () => {
    const { container } = renderTextField()
    expect(container).toMatchSnapshot()
  })
})
