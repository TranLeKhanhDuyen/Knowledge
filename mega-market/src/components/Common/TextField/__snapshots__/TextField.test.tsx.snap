// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField matches snapshot 1`] = `
<div>
  <div
    class="input-container"
  >
    <input
      class="text-field undefined hide"
      placeholder="Enter username"
      type="text"
      value=""
    />
    <p
      class="error-message"
    >
      Username too short
    </p>
  </div>
</div>
`;
